crates/rspack_plugin_split_chunks_new/src/plugin/max_request.rs

```diff
@@ -0,0 +1,67 @@
+use std::borrow::Cow;
+
+use rspack_core::{ChunkUkey, Compilation};
+use rustc_hash::FxHashSet;
+
+use crate::{CacheGroup, SplitChunksPlugin};
+
+impl SplitChunksPlugin {
+  /// Affected by `splitChunks.maxInitialRequests`/`splitChunks.cacheGroups.{cacheGroup}.maxInitialRequests`
+  /// Affected by `splitChunks.maxAsyncRequests`/`splitChunks.cacheGroups.{cacheGroup}.maxAsyncRequests`
+  pub(crate) fn ensure_max_request_fit(
+    &self,
+    compilation: &Compilation,
+    cache_group: &CacheGroup,
+    used_chunks: &mut Cow<FxHashSet<ChunkUkey>>,
+  ) {
+    let chunk_db = &compilation.chunk_by_ukey;
+    let chunk_group_db = &compilation.chunk_group_by_ukey;
+    let invalided_chunks = used_chunks
+      .iter()
+      .map(|c| c.as_ref(chunk_db))
+      .filter_map(|chunk| {
+        let allowed_max_request = if chunk.is_only_initial(chunk_group_db) {
+          cache_group.max_initial_requests
+        } else if chunk.can_be_initial(chunk_group_db) {
+          u32::max(
+            cache_group.max_initial_requests,
+            cache_group.max_async_requests,
+          )
+        } else {
+          cache_group.max_async_requests
+        };
+
+        // `Chunk`s in `used_chunks` are all code-splitting chunk.
+
+        // If a code-splitting chunk is not split by `SplitChunksPlugin`, the number of requests for
+        // the chunk is 1.
+
+        // If the code-splitting chunks is split by `SplitChunksPlugin`, to load the code-splitting chunk
+        // with correct semantics, we need to also load the chunks derive from the the code-splitting chunk.
+
+        // Chunks derive from the the code-splitting chunk is in the same ChunkGroup with the split chunk.
+
+        // So the number of requests is the length of `ChunkGroup#chunks` which belong to the split code-splitting
+        // chunk.
+
+        let actually_requests = chunk
+          .groups
+          .iter()
+          .map(|g| g.as_ref(chunk_group_db))
+          .map(|group| group.chunks.len())
+          .reduce(usize::max)
+          .map(|requests| requests as u32)
+          .unwrap_or_default();
+
+        if actually_requests >= allowed_max_request {
+          Some(chunk.ukey)
+        } else {
+          None
+        }
+      })
+      .collect::<Vec<_>>();
+    invalided_chunks.into_iter().for_each(|c| {
+      used_chunks.to_mut().remove(&c);
```

Comment:
Haven't dig into how webpack implements this feature. Could you describe more about what the relationship between `used_chunks` and `max_xxx_requests` is and how it is affected?

Reply:
I add comments at https://github.com/web-infra-dev/rspack/pull/3119/files/1fa4aa8b4bc5d87ef1121f575b8b92d7d1a843fb#diff-df5e22b8430b27c0d6936f0c0a74a3f449f5b84a3b60d7c84eacb5fb719e88b9R34-R45.

`Chunk`s in `used_chunks` are all code-splitting chunk, which means there are created by entry config or dynamic import.

If a code-splitting chunk is not split by `SplitChunksPlugin`, the number of requests for the chunk is 1.

If the code-splitting chunks is split by `SplitChunksPlugin`, to load the code-splitting chunk, with correct semantics, we need to also load the chunks derive from the the code-splitting chunk. Chunks derive from the the code-splitting chunk is in the same ChunkGroup with the split chunk.

So the number of requests is the length of `ChunkGroup#chunks` which belong to the split code-splitting chunk.

`max_xxx_requests` filter out chunks in `used_chunks` that doesn't satisfy `max_xxx_requests`.

You could see the described behavior at https://webpack.js.org/plugins/split-chunks-plugin/#splitchunksmaxinitialrequests.

Reply:
In general speaking, `max_xxx_requests` controls a code-splitting chunk(create in code splitting) that could be split how many times.

Reply:
I fully understand the semantics of loading a module, however, what I don't understand is how removing `used_chunks` keeps the same semantics and when will the chunks removed from `used_chunks` be loaded.